### PR TITLE
ViewKeeper initial implementation (#47)

### PR DESCRIPTION
**Pull request by warpech**
Implements #47

The architecture is explained in `ViewKeeper.json.cs`.
Based on the Workspaces implementation in https://github.com/StarcounterAppsFreezer/Launcher
Includes two tests.

This is the best I can provide, but frankly speaking, I wish I could implement it as a separate middleware in WebsiteProvider. I just couldn't find a way to do it.

@un-tone please review this thoroughly. If you have any refactor proposals, could you just commit them here? 

Also, please let me know if:

1. The comments that I made in code are enough
2. Should this be explained in README.md?
3. Should this be configurable (turn on/off) in WebsiteProvider settings? In Launcher, workspaces were not configurable

### CODE REVIEW

**Comment by un-tone** [comment]
It really looks as a cool feature!

But currently I'm getting some errors.

#### 1. Case with Signin

1. Run **WebsiteEditor** and **WebsiteProvider** apps from the current branch.
1. Run **SignIn** app from `develop` branch.
1. Go to <http://localhost:8080/signin/createadminuser>.
1. If user is not created do it.
1. Click on "Back" button - nothing is happen.
1. Click on "Back" button again - get the following error:
```
Starcounter.XSON.JsonPatchException: Unknown property in path. Patch: '{"op":"replace","path":"/WebsiteProvider_0/Sections/Main/WebsiteProvider_1/KeptViews/0/OriginalResponse/SignIn_0/Partial/SignIn_0/BackTrigger$","value":"2"}'. Property: 'BackTrigger$'. Viewmodel versions: {client (_ver#c$): 2, clients serverversion: 2, server (_ver#s): 2}.
   в Starcounter.XSON.JsonProperty.EvalutateCurrent(ViewModelVersion version) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonProperty.cs:строка 172
   в Starcounter.Advanced.XSON.JsonExtension.Scope(Json json, Action action) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Advanced\JsonExtension.cs:строка 69
   в Starcounter.XSON.JsonProperty.DoEvaluate(Json root) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonProperty.cs:строка 88
   в Starcounter.XSON.JsonProperty.Evaluate(JsonPointer pointer, Json root) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonProperty.cs:строка 62
   в Starcounter.XSON.DefaultPatchHandler.Handle(Json root, JsonPatchOperation patchOp, JsonPointer pointer, String value) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\DefaultPatchHandler.cs:строка 19
   в Starcounter.XSON.JsonPatch.HandleOnePatch(Json root, JsonPatchOperation op, JsonPointer ptr, String value, Int32 patchCount, ChangeLog changeLog, Object sourceJson, Boolean strictRejection, JsonPatchStatus previousStatus) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonPatch.cs:строка 605
   в Starcounter.XSON.JsonPatch.Apply(Json root, JsonTextReader reader, Boolean strictPatchRejection, Object sourceJson, Int32& patchCount) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonPatch.cs:строка 526
   в Starcounter.XSON.JsonPatch.Apply(Json root, String patch, Boolean strictPatchRejection, Int32& patchCount) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonPatch.cs:строка 414
   в Starcounter.Internal.PuppetRestHandler.HandleWebSocketJson(String bs, WebSocket ws) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.Apps.JsonPatch\BuiltInRestHandlers\PuppetRestHandler.cs:строка 69
Patch={"op":"replace","path":"/WebsiteProvider_0/Sections/Main/WebsiteProvider_1/KeptViews/0/OriginalResponse/SignIn_0/Partial/SignIn_0/BackTrigger$","value":"2"}
CurrentProperty=BackTrigger$
Severity=1
HResult=-2146233088
```

#### 2. Case with WebsiteEditor

1. Run **WebsiteEditor** and **WebsiteProvider** apps from the current branch.
1. Run **SignIn** app from `develop` branch.
1. Go to <http://localhost:8080/websiteeditor> and log in.
1. Click on any surface link and get the following error:

<details>
  <summary>System.NullReferenceException (click to expand)</summary>
System.NullReferenceException: Ссылка на объект не указывает на экземпляр объекта.
   в Starcounter.Templates.Property`1.CheckAndSetBoundValue(Json parent, Boolean addToChangeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Templates\Foundation\Property.cs:строка 220
   в Starcounter.Json.CheckOneTemplate(Json parent, TValue template, ChangeLog changeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 351
   в Starcounter.Json.CollectChanges(ChangeLog changeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 235
   в Starcounter.Json.RetriableCollectChanges(ChangeLog changeLog, Boolean retryOnError) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 224
   в Starcounter.Json.<>c__DisplayClass124_0.<ScopeAndCollectChanges>b__0() в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 198
   в Starcounter.Transaction.Scope(Action action) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter\Transaction.cs:строка 216
   в Starcounter.Advanced.XSON.JsonExtension.Scope(Json json, Action action) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Advanced\JsonExtension.cs:строка 67
   в Starcounter.Json.CheckOneTemplate(Json parent, TValue template, ChangeLog changeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 349
   в Starcounter.Json.CollectChanges(ChangeLog changeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 242
   в Starcounter.Json.RetriableCollectChanges(ChangeLog changeLog, Boolean retryOnError) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 217
   в Starcounter.Json.<>c__DisplayClass124_0.<ScopeAndCollectChanges>b__0() в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 198
   в Starcounter.Internal.TransactionManager.Scope(TransactionHandle handle, Action action) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter\TransactionManager.cs:строка 551
   в Starcounter.Transaction.Scope(Action action) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter\Transaction.cs:строка 224
   в Starcounter.Advanced.XSON.JsonExtension.Scope(Json json, Action action) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Advanced\JsonExtension.cs:строка 67
   в Starcounter.XSON.SiblingList.CollectChanges(ChangeLog changeLog, Json calledFrom) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\SiblingList.cs:строка 270
   в Starcounter.Json.<>c__DisplayClass124_0.<ScopeAndCollectChanges>b__0() в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 201
   в Starcounter.Advanced.XSON.JsonExtension.Scope(Json json, Action action) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Advanced\JsonExtension.cs:строка 69
   в Starcounter.Json.CheckOneTemplate(Json parent, TValue template, ChangeLog changeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 349
   в Starcounter.Json.CollectChanges(ChangeLog changeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 242
   в Starcounter.Json.RetriableCollectChanges(ChangeLog changeLog, Boolean retryOnError) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 217
   в Starcounter.Json.<>c__DisplayClass124_0.<ScopeAndCollectChanges>b__0() в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 198
   в Starcounter.Advanced.XSON.JsonExtension.Scope(Json json, Action action) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Advanced\JsonExtension.cs:строка 69
   в Starcounter.Json.CollectChangesForArray(Json arr, ChangeLog changeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 307
   в Starcounter.Json.CollectChanges(ChangeLog changeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 251
   в Starcounter.Json.RetriableCollectChanges(ChangeLog changeLog, Boolean retryOnError) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 217
   в Starcounter.Json.<>c__DisplayClass124_0.<ScopeAndCollectChanges>b__0() в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 198
   в Starcounter.Advanced.XSON.JsonExtension.Scope(Json json, Action action) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Advanced\JsonExtension.cs:строка 69
   в Starcounter.Json.CheckOneTemplate(Json parent, TValue template, ChangeLog changeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 349
   в Starcounter.Json.CollectChanges(ChangeLog changeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 242
   в Starcounter.Json.RetriableCollectChanges(ChangeLog changeLog, Boolean retryOnError) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 217
   в Starcounter.Json.<>c__DisplayClass124_0.<ScopeAndCollectChanges>b__0() в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 198
   в Starcounter.Advanced.XSON.JsonExtension.Scope(Json json, Action action) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Advanced\JsonExtension.cs:строка 69
   в Starcounter.Json.CheckOneTemplate(Json parent, TValue template, ChangeLog changeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 349
   в Starcounter.Json.CollectChanges(ChangeLog changeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 242
   в Starcounter.Json.RetriableCollectChanges(ChangeLog changeLog, Boolean retryOnError) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 217
   в Starcounter.Json.<>c__DisplayClass124_0.<ScopeAndCollectChanges>b__0() в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 198
   в Starcounter.Advanced.XSON.JsonExtension.Scope(Json json, Action action) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Advanced\JsonExtension.cs:строка 69
   в Starcounter.Json.CheckOneTemplate(Json parent, TValue template, ChangeLog changeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 349
   в Starcounter.Json.CollectChanges(ChangeLog changeLog) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 242
   в Starcounter.Json.RetriableCollectChanges(ChangeLog changeLog, Boolean retryOnError) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 217
   в Starcounter.Json.<>c__DisplayClass124_0.<ScopeAndCollectChanges>b__0() в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\Json.Object.ChangeLog.cs:строка 198
   в Starcounter.Advanced.XSON.JsonExtension.Scope(Json json, Action action) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Advanced\JsonExtension.cs:строка 69
   в Starcounter.XSON.ChangeLog.Generate(Boolean flushLog, Change[]& changeArr) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\ChangeLog.cs:строка 103
   в Starcounter.XSON.JsonPatch.Generate(Json json, Boolean flushLog, Boolean includeNamespace) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonPatch.cs:строка 144
   в Starcounter.XSON.JsonMimeConverter.Convert(Json json, MimeType mimeType, String& resultingMimeType) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.XSON\Class Json\JsonMimeConverter.cs:строка 52
   в Starcounter.Internal.ResourceMimeConverter.TryConvert(Request request, IResource resource, String mimeTypeStr, String& resultingMimeType) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.Apps.JsonPatch\BuiltInRestHandlers\ResourceMimeConverter.cs:строка 67
   в Starcounter.Internal.ResourceMimeConverter.Convert(Request request, Response response, String& resultingMimeType) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.Apps.JsonPatch\BuiltInRestHandlers\ResourceMimeConverter.cs:строка 46
   в Starcounter.Response.ConvertResource(Request req) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.Internal\Http\Response.cs:строка 765
   в Starcounter.Response.ConstructFromFields(Request req, Byte[] givenBuffer) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.Internal\Http\Response.cs:строка 821
   в Starcounter.Request.SendResponse(Response resp, Byte[] serializationBuf) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.Internal\Http\Request.cs:строка 1243
   в Starcounter.Internal.AppsBootstrapper.ProcessExternalRequest(Request req) в C:\TeamCity\BuildAgent\work\sc-8734\Level1\src\Starcounter.Apps.JsonPatch\AppsBootstrapper.cs:строка 394
Data {
IsRethrown=True
}
HResult=-2147467261
</details>

**Comment by un-tone** [comment]
> Should this be explained in README.md?

I think it is not required.

> Should this be configurable

I think there is no need to do so.

File: src/WebsiteProvider/wwwroot/WebsiteProvider/views/ViewKeeper.html [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿<template>
+    <style>
+        .website-surface-cached-active {
+            height: 100%;
+        }
+    </style>
+    <template is="dom-bind">
+        <template is="dom-repeat" items="{{model.KeptViews}}">
+            <template is="dom-if" if="{{item.Active}}">
```
Comment (tomalec) [inline src/WebsiteProvider/wwwroot/WebsiteProvider/views/ViewKeeper.html]: Doesn't `<dom-if>` actually remove elements from DOM?

File: test/WebsiteProvider.Tests/Utilities/OperationObject.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                var hashCode = (Op != null ? Op.GetHashCode() : 0);
+                hashCode = (hashCode * 397) ^ (Path != null ? Path.GetHashCode() : 0);
+                hashCode = (hashCode * 397) ^ (Value != null ? Value.GetHashCode() : 0);
+                return hashCode;
+            }
+        }
+
+        public override string ToString()
+        {
+            return $"Op: {Op}, Path: {Path}, Value: {Value}";
```
Comment (tomalec) [inline test/WebsiteProvider.Tests/Utilities/OperationObject.cs]: Shouldn't we use lover case keys as per RFC?

Comment (warpech) [reply test/WebsiteProvider.Tests/Utilities/OperationObject.cs]: Sorry, I just copied this from a test in Launcher: https://github.com/StarcounterAppsFreezer/Launcher/blame/master/test/Launcher_AcceptanceTest/InactiveWorkspacesAreFrozen.cs#L57

Yes, the generated JSON should have lowercase keys as in the spec: https://tools.ietf.org/html/rfc6902

@un-tone can you pls fix?

File: src/WebsiteProvider/wwwroot/WebsiteProvider/views/ViewKeeper.html (hunk 1 above)
Comment (warpech) [reply src/WebsiteProvider/wwwroot/WebsiteProvider/views/ViewKeeper.html]: https://www.polymer-project.org/1.0/docs/api/elements/dom-if

`dom-if` with `restamp` attribute removes the elements from DOM.

Without that attribute it only hides them with CSS, which makes this PR actually work.

File: test/WebsiteProvider.Tests/Utilities/OperationObject.cs (hunk 2 above)
Comment (un-tone) [reply test/WebsiteProvider.Tests/Utilities/OperationObject.cs]: Fixed.

**Comment by un-tone** [comment]
There are still some issues. I'm currently working on it.

**Comment by un-tone** [comment]
Fixed the issues. Please review the PR.

**Comment by warpech** [comment]
Thanks @PatrykSzwer for the review! I will do the final review today.

**Comment by warpech** [comment]
For some reason, the build for https://github.com/Starcounter/Website/pull/211/commits/0db0237884327355289842bf7101bd129388aef2 hung on TeamCity for more than 1 hour: http://teamcity.starcounter.org/viewLog.html?buildId=34654&tab=buildResultsDiv&buildTypeId=Apps_Website_Develop

**Comment by warpech** [comment]
Something's wrong, because the page http://localhost:8080/WebsiteProvider_AcceptanceHelperOne/ViewKeeperTestSubject does not render the page. As if the `PartialToStandaloneHtmlProvider` middleware was not working correctly. @un-tone any ideas?

**Comment by warpech** [comment]
@un-tone https://github.com/Starcounter/Website/pull/211/commits/88cba20664cc52a158886b98144d4e5bd711c26d looks quite unelegant. Do you remember the problem that it is fixing?

Remaining work:

- [ ] add a Settings page to WebsiteEditor that allows to configure whether ViewKeeper should be used

**Comment by un-tone** [comment]
> Do you remember the problem that it is fixing?

I guess you are about `.ToJson()` using. For example, `response.Resource` can be a `SignIn.MasterPage` instance and `keptView.Response?.Resource` can be the same instance, but it missed the fact that `MasterPage.Partial` was changed. This is the reason why I need to check all content, not only references.

**Comment by warpech** [comment]
Can you pls add a test for that?

**Comment by un-tone** [comment]
> Can you pls add a test for that?

Added `PageLoadTests.GoToSameAppPage_WrappedInMasterPage_Loaded` test.
To check it you can replace the code in ViewKeeper.json.cs with the following code:

```csharp
        public void WrapResponse(Response response)
        {
            KeptViewsElement keptView = this.EnsureKeptView(response.AppName);
            var json = (Json)response.Resource;

            if (keptView.Response?.Resource != json)
            {
                keptView.Response = response;
                keptView.OriginalResponse.MergeJson(json);
            }
        }
```
```csharp
    [ViewKeeper_json.KeptViews]
    partial class KeptViewsElement : Json
    {
        public Response Response;
    }
```

**Comment by un-tone** [comment]
> add a Settings page to WebsiteEditor that allows to configure whether ViewKeeper should be used

Added WebSettings into Simplified: https://github.com/StarcounterApps/Simplified/pull/88.
While I'm waiting when the settings page will be created in scope of https://github.com/Starcounter/Website/issues/251 I created the following temporary URIs for configuring ViewKeeper:
* `/websiteeditor/settings/viewkeeper/enable`
* `/websiteeditor/settings/viewkeeper/disable`

File: src/WebsiteEditor/Api/MainHandlers.cs [hunk 3]
```diff
@@ -66,6 +66,21 @@ public void Register()
...
+        public Response ChangeViewKeeperState(bool enabled)
+        {
+            Db.Transact(() =>
+            {
+                var settings = DataHelper.GetSettings();
+                settings.EnableViewKeeper = enabled;
+            });
+            var state = enabled ? "enabled" : "disabled";
+            Handle.SetOutgoingStatusCode(200);
+            return $"ViewKeeper {state}";
```
Comment (warpech) [inline src/WebsiteEditor/Api/MainHandlers.cs]: Please change the handlers to only work in `localhost`. This is similar how SignIn allows setting the admin user.

Comment (un-tone) [reply src/WebsiteEditor/Api/MainHandlers.cs]: Done.

**Comment by un-tone** [comment]
Added info in the README. I hope it is enough.

File: src/WebsiteEditor/Api/MainHandlers.cs [hunk 4]
```diff
@@ -67,12 +67,18 @@ public void Register()
...
-            Handle.GET("/websiteeditor/settings/viewkeeper/disable", () => this.ChangeViewKeeperState(false));
+            Handle.GET("/websiteeditor/settings/viewkeeper/enable", (Request request) => this.ChangeViewKeeperState(true, request));
+            Handle.GET("/websiteeditor/settings/viewkeeper/disable", (Request request) => this.ChangeViewKeeperState(false, request));
         }
 
-        public Response ChangeViewKeeperState(bool enabled)
+        public Response ChangeViewKeeperState(bool enabled, Request request)
         {
+            var ip = request.ClientIpAddress.ToString();
+            if (ip != "127.0.0.1" && ip != "localhost")
```
Comment (warpech) [inline src/WebsiteEditor/Api/MainHandlers.cs]: `ip != "localhost"` this looks wrong, doesn't `ClientIpAddress` always return an IP?

**Comment by warpech** [comment]
I have extended the info a bit. Did you do anything on TeamCity to fix the build? Now it is passing.

If you agree this is finished now, please merge to `develop`.

**Pull request merged**